### PR TITLE
Support for client-side logging of SOAP calls

### PR DESCRIPTION
**Pull request by sparre-consafe**
This patch allows the user to attach logging call-backs to the generated SOAP client calls.

### CODE REVIEW

File: tools/wsdl2aws-generator-stub.adb [hunk 1]
```diff
@@ -488,6 +496,14 @@ package body Stub is
          Text_IO.New_Line (Stub_Adb);
       end if;
 
+      if O.Gen_CB then
+         Text_IO.Put_Line
+            (Stub_Adb,
+             "      Post_Call_Callback (Connection, " &
+                """" & To_String (O.Prefix) & Proc & """, " &
+                "Payload, Response, Schema);");
```
Comment (TurboGit) [inline tools/wsdl2aws-generator-stub.adb]: note that Schema is a global variable already available to users. I would not pass it here.

File: src/soap/soap-client-callback.ads [hunk 2]
```diff
@@ -0,0 +1,34 @@
+with SOAP.Message.Payload;
+with SOAP.Message.Response;
+
+package SOAP.Client.Callback is
+
+   --  Callback routine types
+   type Pre_Call_CB is not null access
+      procedure (Connection : AWS.Client.HTTP_Connection;
+                 SOAPAction : String;
```
Comment (TurboGit) [inline src/soap/soap-client-callback.ads]: This SOAPAction is part of the Request parameter. Not sure it is needed in the CB, one can get it with

   SOAPAction : constant String := Status.SOAPAction (Request);

File: tools/wsdl2aws-generator-stub.adb (hunk 1 above)
Comment (Knaldgas) [reply tools/wsdl2aws-generator-stub.adb]: Can I just use SOAP.WSDL.Schema.Definition?
The callback might not know which request caused the callback.

Comment (TurboGit) [reply tools/wsdl2aws-generator-stub.adb]: No, the Schema variable is generated at the top of the package hierarchy. See root package of a generated code. And, BTW, the Schema variable above passed to Post_Call_Callback is exactly this variable. Note also that the Schema variable contains the definitions for the whole Web Service and is never modified after the initialisation phase (elaboration).

Comment (Knaldgas) [reply tools/wsdl2aws-generator-stub.adb]: I see a Schema definition in the top of the package hierarchy of a specific service-set only. We have multiple service-sets aggregated with awsascb, but a common handler for callbacks across these service-sets.

Comment (TurboGit) [reply tools/wsdl2aws-generator-stub.adb]: Indeed, I was thinking about this. So let's keep Schema in the callback.

File: tools/wsdl2aws-generator-stub.adb [hunk 3]
```diff
@@ -488,6 +496,14 @@ package body Stub is
          Text_IO.New_Line (Stub_Adb);
       end if;
 
+      if O.Gen_CB then
```
Comment (TurboGit) [inline tools/wsdl2aws-generator-stub.adb]: I'm not sure we want to have this feature tied to the -cb option. We can probably want to be able to get traces without having generated the the SOAP callbacks. The traces are in the client part of the code. Maybe a new option -traces would be better. Do you agree?

File: src/soap/soap-client-callback.ads (hunk 2 above)
Comment (sparre-consafe) [reply src/soap/soap-client-callback.ads]: Probably better to do it like that then.

File: tools/wsdl2aws-generator-stub.adb (hunk 3 above)
Comment (sparre-consafe) [reply tools/wsdl2aws-generator-stub.adb]: I agree, but was too lazy to figure out how to add a new option.  I'll take a look at it.

**Comment by sparre-consafe** [comment]
I'll give it a new try from a dedicated branch.

**Pull request closed without merging**